src/lib/draw-and-pick.js

```diff
@@ -69,17 +75,39 @@ export function pickLayers(gl, {
   const deviceX = x * pixelRatio;
   const deviceY = gl.canvas.height - y * pixelRatio;
 
-  // TODO - just return glContextWithState once luma updates
   const unhandledPickInfos = [];
+  let pickedColor;
+  let pickedLayer;
+  let pickedObjectIndex;
+  const affectedLayers = [];
 
-  // Make sure we clear scissor test and fbo bindings in case of exceptions
-  // We are only interested in one pixel, no need to render anything else
-  glContextWithState(gl, {
-    frameBuffer: pickingFBO,
-    framebuffer: pickingFBO,
-    scissorTest: {x: deviceX, y: deviceY, w: 1, h: 1}
-  }, () => {
+  if (MOTION_EVENTS.indexOf(mode) !== -1) {
+    // "Motion events" are those that track the motion of an interaction
+    // after its initiation. In this case, these subsequent events
+    // should be bound to the object that was first picked,
+    // e.g. to enable dragging behaviors.
+    // Therefore, the picking process does not run for these events.
+    const {layerId} = lastPickedInfo;
+    pickedLayer = layers.find(l => l.props.id === layerId);
+    if (pickedLayer) {
+      pickedColor = lastPickedInfo.color;
+      pickedObjectIndex = lastPickedInfo.index;
+      affectedLayers.push(pickedLayer);
+    }
+  } else {
+    // Make sure we clear scissor test and fbo bindings in case of exceptions
+    // We are only interested in one pixel, no need to render anything else
+    // Note that glContextCallback is called synchronously;
+    // code in this function is executed in the order it's written.
+    // TODO - just return glContextWithState once luma updates
+    glContextWithState(gl, {
```

Comment:
While I agree that this function need to be split into smaller functions, I think this is split in the wrong place. Move the `glContextWithState` into the sub function and rename it something meaningful.

It will make the diff a little smaller also since indentation will be unaffected. It would help my review if this was the case.

Just in case: what `glContextWithState` does is that it locally sets some values on the gl context and guarantees that they are reset even if we exit with an exception.